### ISSUE TITLE
Support unfold for Long tensors on CPU

### ISSUE THREAD

**Issue by vadimkantorov** [pytorch/pytorch#44989]
It would be useful for mode filter that should operate on discrete inputs:
``` 
import torch

def mode_pooling1d(tensor, kernel_size, stride, padding):
    return torch.nn.functional.unfold(tensor[:, None, None, :], kernel_size = (1, kernel_size), padding = (0, padding), stride = (1, stride)).mode(1).values

if __name__ == '__main__':
    x = torch.randint(low = 0, high = 3, size = (2, 128))
    y = mode_pooling1d(x, kernel_size = 7, padding = 7 // 2, stride = 1)
    print(x.shape, y.shape)
```

**Comment by agolynski** [comment]
@vadimkantorov Seems like unfold is only supported on floating and complex types.

Could you file mode() request in a different issue?

**Comment by vadimkantorov** [comment]
Already done :) https://github.com/pytorch/pytorch/issues/45009

**Comment by vadimkantorov** [comment]
Support of integral tensors is also very important for working on index tensors (which are int64). Applying unfold on an index tensor can then represent indices of neighborhood elements, which can be helpful for local attention implementations for large seqlens

cc @albanD

**Comment by albanD** [comment]
Well, it works by re-interpret cast the content into float64, do the shuffling and then cast it back.
I'm fine with adding this trick to the python implementation in nn/functional.py to enable this.

**Comment by vadimkantorov** [comment]
For the under-the hood version, maybe would still be good to add uint8/float8/float16. Maybe this asks for some opaque dtypes of different widths as proposed in https://github.com/pytorch/pytorch/issues/104702

**Comment by albanD** [comment]
We could add that trick at the kernel level.
Or if we want to be extra fancy, convert all of these to the appropriate `bitsX` dtype and only have implementation for these dtypes :) 
Not sure it's worth the trouble though compared to a 4 lines change in python :D

**Comment by vadimkantorov** [comment]
I think it's worth as uint8/float8 is likely not currently supported now...

**Comment by vadimkantorov** [comment]
This could be a basis for other holistic ops, such as fill / F.pad: https://github.com/pytorch/pytorch/issues/108226
